node/src/Pos/Block/Logic/Header.hs

```diff
@@ -164,8 +172,15 @@ classifyHeaders inRecovery headers = do
     tipHeader <- DB.getTipHeader @ssc
     let tip = headerHash tipHeader
     haveOldestParent <- isJust <$> DB.blkGetHeader @ssc oldestParentHash
-    let headersValid = isVerSuccess $
-                       verifyHeaders (headers & _Wrapped %~ toList)
+    leaders <-
+        lrcActionOnEpochReason
+            newestHeaderEpoch
+            (sformat ("classifyHeaders: there are no leaders for epoch " %build)
+                     newestHeaderEpoch)
+            LrcDB.getLeaders
+    let headersValid =
+            isVerSuccess $
+            verifyHeaders (Just leaders) (headers & _Wrapped %~ toList)
```

Comment:
1. Is it ok that we don't use `dlgVerifyHeader` in this function?
2. Is it guaranteed that all headers are from the same epoch? If not, then passing only one `SlotLeaders` argument can't work I suppose.

Reply:
I see that `verifyHeaders` will use given leaders for given blocks until it reaches `GenesisBlock`. But you pass `newestHeader ^. epochIndexL`. Probably you should at least change it to `oldestHeader ^. epochIndexL`. But it's probably better to pass leaders for all epochs.
P. S. Maybe I just don't know something and blocks are guaranteed to be from the same epoch here.

Reply:
1. It's okay, since we anyway can't verify proxy signatures if we have >1 header. See PR comment.
2. No, it is not guaranteed. Yes, there is no total safety here (and can't be).

`newest ~ oldest` is a bug, thank you. There is no point in passing "leaders for all epochs" because it's not possible (I'm not sure though, maybe it will work for [9k,10k] period, when do we compute lrc?) to know leaders of future epoch in advance. Note: `classifyHeaders` is used in recovery mode now (since default mode is incremental and we check only one header).

Reply:
> There is no point in passing "leaders for all epochs" because it's not possible

On, well, yes. We should also ensure that lrc will be used before you call `lrcActionOnEpochReason`, otherwise it will just hang. I will check it.